src/operator/join_range.rs

```diff
@@ -0,0 +1,294 @@
+    /// receives at each timestamp ignoring previous inputs.
+    pub fn stream_join_range_indexed<RF, JF, It, K, V, I2>(
+        &self,
+        other: &Stream<Circuit<P>, I2>,
+        range_func: RF,
+        join_func: JF,
+    ) -> Stream<Circuit<P>, OrdIndexedZSet<K, V, I1::R>>
+    where
+        I1: BatchReader<Time = ()> + Clone + 'static,
+        I1::R: MulByRef,
+        I2: BatchReader<Time = (), R = I1::R> + Clone + 'static,
+        I2::Key: Ord,
+        RF: Fn(&I1::Key) -> (I2::Key, I2::Key) + 'static,
+        JF: Fn(&I1::Key, &I1::Val, &I2::Key, &I2::Val) -> It + 'static,
+        K: Clone + Ord + 'static,
+        V: Clone + Ord + 'static,
+        It: IntoIterator<Item = (K, V)> + 'static,
+    {
+        self.stream_join_range_generic(other, range_func, join_func)
+    }
+
+    /// Like [`Self::stream_join_range`], but can return any indexed Z-set type.
+    pub fn stream_join_range_generic<RF, JF, It, I2, O>(
+        &self,
+        other: &Stream<Circuit<P>, I2>,
+        range_func: RF,
+        join_func: JF,
+    ) -> Stream<Circuit<P>, O>
+    where
+        I1: BatchReader<Time = (), R = O::R> + Clone + 'static,
+        I2: BatchReader<Time = (), R = O::R> + Clone + 'static,
+        I2::Key: Ord,
+        O: Batch<Time = ()> + Clone + 'static,
+        O::R: MulByRef,
+        RF: Fn(&I1::Key) -> (I2::Key, I2::Key) + 'static,
+        JF: Fn(&I1::Key, &I1::Val, &I2::Key, &I2::Val) -> It + 'static,
+        It: IntoIterator<Item = (O::Key, O::Val)> + 'static,
+    {
+        self.circuit()
+            .add_binary_operator(StreamJoinRange::new(range_func, join_func), self, other)
+    }
+}
+
+pub struct StreamJoinRange<RF, JF, It, I1, I2, O> {
+    range_func: RF,
+    join_func: JF,
+    _types: PhantomData<(It, I1, I2, O)>,
+}
+
+impl<RF, JF, It, I1, I2, O> StreamJoinRange<RF, JF, It, I1, I2, O> {
+    pub fn new(range_func: RF, join_func: JF) -> Self {
+        Self {
+            range_func,
+            join_func,
+            _types: PhantomData,
+        }
+    }
+}
+
+impl<RF, JF, It, I1, I2, O> Operator for StreamJoinRange<RF, JF, It, I1, I2, O>
+where
+    I1: 'static,
+    I2: 'static,
+    It: 'static,
+    RF: 'static,
+    JF: 'static,
+    O: 'static,
+{
+    fn name(&self) -> Cow<'static, str> {
+        Cow::from("StreamJoinRange")
+    }
+    fn fixedpoint(&self, _scope: Scope) -> bool {
+        true
+    }
+}
+
+impl<RF, JF, It, I1, I2, O> BinaryOperator<I1, I2, O> for StreamJoinRange<RF, JF, It, I1, I2, O>
+where
+    I1: BatchReader<Time = (), R = O::R> + Clone + 'static,
+    I2: BatchReader<Time = (), R = O::R> + Clone + 'static,
+    I2::Key: Ord,
+    O: Batch<Time = ()> + 'static,
+    O::R: MulByRef,
+    RF: Fn(&I1::Key) -> (I2::Key, I2::Key) + 'static,
+    JF: Fn(&I1::Key, &I1::Val, &I2::Key, &I2::Val) -> It + 'static,
+    It: IntoIterator<Item = (O::Key, O::Val)> + 'static,
+{
+    fn eval(&mut self, i1: &I1, i2: &I2) -> O {
```

Comment:
Some comments/docs on what's doing on within this would be good for future readers